tsl/test/src/remote/node_killer.c

```diff
@@ -17,88 +17,142 @@
 #include "export.h"
 #include "remote/connection.h"
 #include "remote/dist_txn.h"
+
 #include "test_utils.h"
 #include "connection.h"
 
-typedef struct RemoteNodeKiller
+static RemoteNodeKiller rnk_event = {
+	.conn = NULL,
+	.pid = -1,
+	.killevent = DTXN_EVENT_ANY,
+};
+
+static void
+kill_backend(const DistTransactionEvent event, void *data)
 {
-	pid_t pid;
-	TSConnection *conn;
-} RemoteNodeKiller;
+	RemoteNodeKiller *rnk = data;
 
-static char *kill_event = NULL;
-static RemoteNodeKiller *rnk_event = NULL;
+	if (event == rnk->killevent || rnk->killevent == DTXN_EVENT_ANY)
+	{
+		elog(WARNING, "kill event: %s", remote_dist_txn_event_name(rnk->killevent));
+		remote_node_killer_kill(&rnk_event);
+		remote_dist_txn_set_event_handler(NULL);
+	}
+}
+
+static DistTransactionEventHandler eventhandler = {
+	.handler = kill_backend,
+	.data = &rnk_event,
+};
 
 TS_FUNCTION_INFO_V1(ts_remote_node_killer_set_event);
 
-RemoteNodeKiller *
-remote_node_killer_init(TSConnection *conn)
+void
+remote_node_killer_init(RemoteNodeKiller *rnk, const TSConnection *conn,
+						const DistTransactionEvent event)
 {
-	RemoteNodeKiller *rnk = palloc(sizeof(RemoteNodeKiller));
+	int remote_pid = remote_connecton_get_remote_pid(conn);
+
+	if (remote_pid == -1)
+		elog(ERROR, "could not get PID of remote backend process");
 
+	MemSet(rnk, 0, sizeof(*rnk));
```

Comment:
nit: Is this PostgreSQL version?  why not using regular `memset()`?

Reply:
yeah, PostgreSQL version. Apparently faster in some cases. I don't think it matters what we use here since it is not on a critical path.